ec2/utils.py

```diff
@@ -76,120 +215,267 @@ def get_instance_state(instance, ctx):
+    except boto.exception.EC2ResponseError as e:
+        group = get_security_group_from_name(group_id, ctx=ctx)
+        raise NonRecoverableError('Error. '
+                                  'Failed to group by id: '
+                                  'API returned: {0}.'
+                                  .format(str(e)))
+    except boto.exception.BotoServerError as e:
+        raise NonRecoverableError('Error. '
+                                  'Failed to group by id: '
+                                  'API returned: {0}.'
+                                  .format(str(e)))
 
-def validate_group(group, ctx):
-    ctx.logger.debug('Testing if group with identifier '
-                     ' {0} exists in this account.'.format(group))
-    groups = get_security_group_from_id(group, ctx)
+    if len(group) < 1:
+        group = get_security_group_from_name(group_id, ctx=ctx)
+    elif len(group) > 1:
+        report_all_security_groups(ctx=ctx)
+        raise NonRecoverableError('Error. '
+                                  'Failed to group by id or name: '
+                                  'Too many groups returned.')
 
-    if groups is not None:
-        return True
-    else:
-        return False
+    return group
 
 
-def get_security_group_from_id(group, ctx):
+def get_security_group_from_name(group_name, ctx):
     ec2_client = connection.EC2ConnectionClient().client()
-    ctx.logger.debug('Getting Security Group by ID: {0}'.format(group))
-
     try:
-        groups = ec2_client.get_all_security_groups(group_ids=group)
-    except (EC2ResponseError, BotoServerError) as e:
-        raise NonRecoverableError('(Node: {0}): Error. '
+        group = ec2_client.get_all_security_groups(groupnames=group_name)
+    except boto.exception.EC2ResponseError as e:
+        report_all_security_groups(ctx=ctx)
+        raise NonRecoverableError('Error. '
+                                  'Failed to group by id: '
+                                  'API returned: {0}.'
+                                  .format(str(e)))
+    except boto.exception.BotoServerError as e:
+        raise NonRecoverableError('Error. '
                                   'Failed to group by id: '
-                                  'API returned: {1}.'
-                                  .format(ctx.instance.id, e))
-    return groups
+                                  'API returned: {0}.'
+                                  .format(str(e)))
+    if len(group) is not 1:
+        report_all_security_groups(ctx=ctx)
+        raise NonRecoverableError('Error. '
+                                  'Failed to group by id or name: '
+                                  'Too many groups returned.')
 
+    return group
 
-def save_key_pair(key_pair_object, ctx):
-    """ Saves the key pair to the file specified in the blueprint
-    """
 
-    ctx.logger.debug('Attempting to save the key_pair_object.')
+def get_key_pair_by_id(ctx):
```

Comment:
We're creating connection clients all the time... maybe we should have the developer declare the client once and then send it to every function they want to execute.